lib/rspec/core/bisect/example_minimizer.rb

```diff
@@ -18,20 +16,52 @@ def initialize(runner, reporter)
         def find_minimal_repro
           prep
 
-          self.remaining_ids = non_failing_example_ids
+          self.remaining_ids, duration = track_duration do
+            bisect_over(non_failing_example_ids)
+          end
 
-          each_bisect_round do |subsets|
-            ids_to_ignore = subsets.find do |ids|
-              get_expected_failures_for?(remaining_ids - ids)
-            end
+          notify(:bisect_complete, :round => 10, :duration => duration,
+                                   :original_non_failing_count => non_failing_example_ids.size,
+                                   :remaining_count => remaining_ids.size)
+
+          remaining_ids + failed_example_ids
+        end
 
-            next :done unless ids_to_ignore
+        def bisect_over(candidate_ids)
+          return [] if get_expected_failures_for?([])
 
-            self.remaining_ids -= ids_to_ignore
-            notify(:bisect_ignoring_ids, :ids_to_ignore => ids_to_ignore, :remaining_ids => remaining_ids)
+          bisect_over_recursive(candidate_ids, [])
+        end
+
+        def bisect_over_recursive(candidate_ids, fixed)
+          return candidate_ids if candidate_ids.length == 1
+          slice_size = (candidate_ids.length / 2.0).ceil
+          slices = candidate_ids.each_slice(slice_size).to_a
+
+          notify(
+            :bisect_round_started,
+            :round => 10,
+            :subset_size => slice_size,
+            :remaining_count => candidate_ids.size
+          )
+
+          ids_to_ignore = slices.find do |ids|
+            get_expected_failures_for?(candidate_ids - ids + fixed)
           end
 
-          currently_needed_ids
+          if ids_to_ignore
+            self.remaining_ids = candidate_ids - ids_to_ignore
```

Comment:
Yeah - I tried to eliminate it but ended up having problems with `#repro_command_for_currently_needed_ids`. I think to support the graceful abort on SIGINT feature the minimizer will have to store its current progress (since the execution state will be lost on interrupt), so I don't think it's going to be easy to get away from this bit of state.

Maybe it'd be good to write the algorithm purely with local variables, then add explicit `#memoize_progress` calls to write the state, just to make it a bit clearer why the object-level state is needed. I'll give that a go and see if it's any nicer.